platform/default/online_file_source.cpp

```diff
@@ -176,6 +167,66 @@ class OnlineFileSource::Impl {
         }
     }
 
+    struct PendingRequests {
+        // initialize two pending queues: a high priority queue for "regular" requests and a low priority queue for offline download requests
+        // the high priority requests are in prioritizedPendingRequests*[0]
+        // the low priority requests are in prioritizedPendingRequests*[1]
+        // context: https://github.com/mapbox/mapbox-gl-native/issues/12655
+        PendingRequests() : prioritizedPendingRequestsList(2), prioritizedPendingRequestsMap(2) {}
+
+        std::vector<std::list<OnlineFileRequest*>> prioritizedPendingRequestsList;
+        std::vector<std::unordered_map<OnlineFileRequest*, std::list<OnlineFileRequest*>::iterator>> prioritizedPendingRequestsMap;
+
+        void remove(OnlineFileRequest* request) {
+            auto it = prioritizedPendingRequestsMap[0].find(request);
+            if (it != prioritizedPendingRequestsMap[0].end()) {
+                prioritizedPendingRequestsList[0].erase(it->second);
+                prioritizedPendingRequestsMap[0].erase(it);
+            }
+            else {
+                it = prioritizedPendingRequestsMap[1].find(request);
+                if (it != prioritizedPendingRequestsMap[1].end()) {
+                    prioritizedPendingRequestsList[1].erase(it->second);
+                    prioritizedPendingRequestsMap[1].erase(it);
+                }
+            }
+            assertSizes();
+        }
+
+        void insert(OnlineFileRequest* request) {
+            std::size_t priority = request->resource.isLowPriority;
+            auto it = prioritizedPendingRequestsList[priority].insert(prioritizedPendingRequestsList[priority].end(), request);
+            prioritizedPendingRequestsMap[priority].emplace(request, std::move(it));
+            assertSizes();
+        }
+
+        bool empty() {
+            return prioritizedPendingRequestsList[0].empty() && prioritizedPendingRequestsList[1].empty();
+        }
+
+        OnlineFileRequest* pop() {
+            assert(!empty());
+            std::size_t next_priority = !prioritizedPendingRequestsList[0].empty() ? 0 : 1;
+
+            OnlineFileRequest* request = prioritizedPendingRequestsList[next_priority].front();
+            prioritizedPendingRequestsList[next_priority].pop_front();
+            prioritizedPendingRequestsMap[next_priority].erase(request);
+            return request;
+
+            assertSizes();
+        }
```

Comment:
We're never checking `empty()` without also `pop`ping a request. Instead of having a separate check for emptiness and a pop operation, you could return a null pointer when the request list is empty. This avoids the possibility of the two functions becoming out of sync once function calls are moved around.

Reply:
We could also combine the prioritized and the regular list by inserting new priorities item at the end of the prioritized item range (== in the "middle" of the list), rather than inserting them in the end and remove the need for two data structures, as well as additional logic when popping items.